extensions/hashicorp-vault/src/main/java/org/eclipse/dataspaceconnector/core/security/hashicorpvault/HashicorpVaultExtension.java

```diff
@@ -0,0 +1,96 @@
+/*
+ *  Copyright (c) 2022 Mercedes-Benz Tech Innovation GmbH
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Mercedes-Benz Tech Innovation GmbH - Initial API and Implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.core.security.hashicorpvault;
+
+import dev.failsafe.RetryPolicy;
+import okhttp3.OkHttpClient;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.EdcSetting;
+import org.eclipse.dataspaceconnector.spi.security.PrivateKeyResolver;
+import org.eclipse.dataspaceconnector.spi.security.Vault;
+import org.eclipse.dataspaceconnector.spi.security.VaultPrivateKeyResolver;
+import org.eclipse.dataspaceconnector.spi.system.Provides;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtension;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtensionContext;
+
+import java.time.temporal.ChronoUnit;
+
+@Provides({ Vault.class, PrivateKeyResolver.class })
+public class HashicorpVaultExtension implements ServiceExtension {
+
+    @EdcSetting(required = true)
+    public static final String VAULT_URL = "edc.vault.hashicorp.url";
+
+    @EdcSetting(required = true)
+    public static final String VAULT_TOKEN = "edc.vault.hashicorp.token";
+
+    @EdcSetting
+    public static final String MAX_RETRIES = "edc.core.retry.retries.max";
+
+    @EdcSetting
+    public static final String BACKOFF_MIN_MILLIS = "edc.core.retry.backoff.min";
+
+    @EdcSetting
+    public static final String BACKOFF_MAX_MILLIS = "edc.core.retry.backoff.max";
+
+    @Override
+    public String name() {
+        return "Hashicorp Vault";
+    }
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        var config = loadHashicorpVaultClientConfig(context);
+
+        var maxRetries = context.getSetting(MAX_RETRIES, 5);
+        var minBackoff = context.getSetting(BACKOFF_MIN_MILLIS, 500);
+        var maxBackoff = context.getSetting(BACKOFF_MAX_MILLIS, 10_000);
+
+        var retryPolicy = RetryPolicy.builder()
+                .withMaxRetries(maxRetries)
+                .withBackoff(minBackoff, maxBackoff, ChronoUnit.MILLIS)
+                .build();
+
+        var okHttpClient = new OkHttpClient.Builder().build();
+        var client = new HashicorpVaultClient(config, okHttpClient, context.getTypeManager(), retryPolicy);
+
+        var vault = new HashicorpVault(client, context.getMonitor());
+        var privateKeyResolver = new VaultPrivateKeyResolver(vault);
+
+        context.registerService(Vault.class, vault);
```

Comment:
registering services could be done using the `@Provider` method functionality, which is the preferred and recommended way. In that case, the `@Provides` annotation on the class can go away.

Reply:
almost there. the `@Provides` annotation on the extension class should be removed as well.